pkg/cloudprovider/providers/openstack/openstack_instances.go

```diff
@@ -54,6 +55,11 @@ func (os *OpenStack) Instances() (cloudprovider.Instances, bool) {
 		return nil, false
 	}
 
+	// Server flavor details (flavor.original_name) requires
+	// micro-version 2.47. Prior to this, only flavor.id is
+	// returned, which is less useful as a node label.
+	compute.Microversion = "2.47"
```

Comment:
This change brings incompatible change to the openstack cloud which doesn't support compute service version 2.47.

Reply:
That's true. Do you think it should fallback to using `flavor.id` if this micro-version is not available? What is the strategy for the other services that also set a micro version?

Reply:
You can try to fetch supported microversions before setting it directly. I guess you can use this package https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/apiversions/requests.go against compute API, because urls are the same https://docs.openstack.org/api-guide/compute/microversions.html

See links below with similar issues:
https://github.com/kubernetes/cloud-provider-openstack/pull/786/files
https://github.com/kubernetes/cloud-provider-openstack/pull/808/files
Also be aware about the shared vars and threads.

Reply:
If the original requirement is to get a human-friendly flavor name, why not just change the order of keys in `srvInstanceType` function?

Reply:
@lingxiankong sorry, that wasn't clear... The problem is that `id` and `original_name` are mutually exclusive in the response and depend on the micro-version selection. As such, changing the order does not make any difference.

Basically, you get either `id` _or_ `original_name`, and this depends on the negotiated micro-version. At and above 2.47, you get `original_name`, and below, you get `id`.

On our Openstack installation, the default micro version (if not specified) is less than 2.47, even though the compute service supports micro-versions up to 2.60.

I will try to make this clearer in the code and PR.

Reply:
the way send to network API is doable but it's a little bit hacking..
can we create a followup PR to gophercloud to add microversion negoation to compute side?

Reply:
Before supporting micro-version negotiation in gophercloud, we need something like https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/util/openstack/loadbalancer.go#L77